test/SILOptimizer/string_switch.swift

```diff
@@ -0,0 +1,101 @@
+
+// RUN: %empty-directory(%t)
+// RUN: %target-build-swift -O -Xllvm -sil-disable-pass=FunctionSignatureOpts -module-name=test %s -o %t/a.out
+// RUN: %target-run %t/a.out | %FileCheck %s -check-prefix=CHECK-OUTPUT
+
+import Swift
+
+enum Dummy : String {
+  case a = "a"
+  case b = "b"
+  case c = "c"
+}
+
+
+// CHECK-LABEL: sil_global private {{.*}}test1{{.*}}v_{{.*}}
+// CHECK: string_literal utf8 "a"
+
+// CHECK-LABEL: sil_global private {{.*}}test1{{.*}}v0_{{.*}}
+// CHECK: string_literal utf8 "b"
+// CHECK: string_literal utf8 "x"
+
+// CHECK-LABEL: sil {{.*}}test0{{.*}}
+func test0(str: String) -> Int {
+  // TODO: check that this is the *only* apply of FN
+  // CHECK: [[FN:%[0-9]+]] = function_ref @${{.*}}_findStringSwitchCase{{.*}}
+  // CHECK: [[CMP:%[0-9]+]] = apply [[FN]]{{.*}}
+  // CHECK: [[EXT:%[0-9]+]] = struct_extract [[CMP]] {{.*}}
+  // CHECK: [[ICMP:%[0-9]+]] = builtin "cmp_eq_Int64"([[EXT]] {{.*}}
+  // CHECK: cond_br [[ICMP]], {{.*}}
+  switch str {
+    case "0": return 0
+    case "2": return 1
+    case "3": return 2
+    default: return -1
+  }
+}
+
+// CHECK-LABEL: sil {{.*}}test1{{.*}}
+func test1(x: Int, input: String) -> Int {
+  // CHECK: [[ONE:%[0-9]+]] = integer_literal $Builtin.Int{{[0-9]+}}, 1
+  // CHECK: [[CMP_ONE:%[0-9]+]] = builtin "cmp_eq_Int64"(%{{[0-9]+}} : $Builtin.Int64, [[ONE]] : $Builtin.Int64)
+  // CHECK: cond_br [[CMP_ONE]], {{.*}}
+  if x == 1 {
+    // CHECK: %{{[0-9]+}} = global_value {{.*}}test1{{.*}}v_{{.*}}
+    // CHECK: [[FN:%[0-9]+]] = function_ref @${{.*}}_findStringSwitchCase{{.*}}
+    // CHECK: [[CMP:%[0-9]+]] = apply [[FN]]{{.*}}
+    // CHECK: [[EXT:%[0-9]+]] = struct_extract [[CMP]] {{.*}}
+    // CHECK: [[ICMP:%[0-9]+]] = builtin "cmp_eq_Int64"([[EXT]] {{.*}}
+    // CHECK: cond_br [[ICMP]], {{.*}}
+    if input == "a" { return 1 }
+    return 2
+  } else {
+    // CHECK: %{{[0-9]+}} = global_value {{.*}}test1{{.*}}v0_{{.*}}
+    // CHECK: [[FN:%[0-9]+]] = function_ref @${{.*}}_findStringSwitchCase{{.*}}
+    // CHECK: [[CMP:%[0-9]+]] = apply [[FN]]{{.*}}
+    // CHECK: [[EXT:%[0-9]+]] = struct_extract [[CMP]] {{.*}}
+    // CHECK: [[ICMP:%[0-9]+]] = builtin "cmp_eq_Int64"([[EXT]] {{.*}}
+    // CHECK: cond_br [[ICMP]], {{.*}}
+    if input == "x" { return x}
+      // CHECK: [[ICMP:%[0-9]+]] = builtin "cmp_eq_Int64"([[EXT]] {{.*}}
+      // CHECK: cond_br [[ICMP]], {{.*}}
+    else if input == "b" { return 2 }
+    return 0
+  }
+}
+
+// CHECK-LABEL: sil {{.*}}test2{{.*}}
+func test2(x: [String]) -> Int {
+  for str in x {
+    // CHECK: [[FN:%[0-9]+]] = function_ref @${{.*}}_findStringSwitchCase{{.*}}
+    // CHECK: [[CMP:%[0-9]+]] = apply [[FN]]{{.*}}
+    // CHECK: [[EXT:%[0-9]+]] = struct_extract [[CMP]] {{.*}}
+    // CHECK: [[ICMP:%[0-9]+]] = builtin "cmp_eq_Int64"([[EXT]] {{.*}}
+    // CHECK: cond_br [[ICMP]], {{.*}}
+    if str == "foo" { return 0 }
```

Comment:
This test reminds me that it you probably should add a limit on the minim number of string compares.
I don't believe that a _findStringSwitchCase is more efficient than a single string compare, both for performance and code size. You can find out by experiment what a good lower bound can be.

Reply:
Your right. Will do.

Reply:
Looks like it is beneficial after one comparison (two and on). Heres a simple graph:
<img width="926" alt="Screen Shot 2019-12-02 at 1 31 30 PM" src="https://user-images.githubusercontent.com/13806068/69997148-5a78a880-1508-11ea-863b-377a65386311.png">